From c2453b121c0610e8e47333c109e40399806a35c8 Mon Sep 17 00:00:00 2001
From: Yunqiang Hu <yunhu@qti.qualcomm.com>
Date: Thu, 14 May 2026 15:46:14 +0800
Subject: [PATCH] event: use memcmp for change detection

Description: Replace byte-by-byte loop with memcmp to reduce CPU overhead.
Signed-off-by: Yunqiang Hu <yunhu@qti.qualcomm.com>
---
 implementation/routing/src/event.cpp | 21 ++++++++++-----------
 1 file changed, 10 insertions(+), 11 deletions(-)

diff --git a/implementation/routing/src/event.cpp b/implementation/routing/src/event.cpp
index 48af1dfb..80c570bb 100644
--- a/implementation/routing/src/event.cpp
+++ b/implementation/routing/src/event.cpp
@@ -4,6 +4,7 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 #include <chrono>
+#include <cstring>
 #include <iomanip>
 #include <sstream>
 #include <thread>
@@ -814,17 +815,15 @@ bool
 event::has_changed(const std::shared_ptr<payload> &_lhs,
         const std::shared_ptr<payload> &_rhs) const {
 
-    bool is_change = (_lhs->get_length() != _rhs->get_length());
-    if (!is_change) {
-        std::size_t its_pos = 0;
-        const byte_t *its_old_data = _lhs->get_data();
-        const byte_t *its_new_data = _rhs->get_data();
-        while (!is_change && its_pos < _lhs->get_length()) {
-            is_change = (*its_old_data++ != *its_new_data++);
-            its_pos++;
-        }
-    }
-    return is_change;
+    if (_lhs->get_length() != _rhs->get_length())
+        return true;
+    const auto its_length = _lhs->get_length();
+    if (its_length == 0)
+        return false;
+    return std::memcmp(
+            _lhs->get_data(),
+            _rhs->get_data(),
+            its_length) != 0;
 }
 
 std::set<client_t>
-- 
2.34.1

